frame/vamm/src/lib.rs

```diff
@@ -0,0 +1,332 @@
+		/// Event type emitted by this pallet. Depends on the runtime's definition of an event.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// The `VammId` used by the pallet. Corresponds to the Ids used by the Vamm pallet.
+		type VammId: Default
+			+ CheckedAdd
+			+ Clone
+			+ Copy
+			+ Debug
+			+ FullCodec
+			+ MaxEncodedLen
+			+ MaybeSerializeDeserialize
+			+ One
+			+ Parameter
+			+ PartialEq
+			+ TypeInfo
+			+ Zero;
+
+		/// Timestamp to be used for twap calculations and market deprecation.
+		type Timestamp: Default
+			+ Clone
+			+ Copy
+			+ Debug
+			+ FullCodec
+			+ MaxEncodedLen
+			+ PartialEq
+			+ TypeInfo;
+
+		/// The Balance type used by the pallet for bookkeeping. `Config::Convert` is used for
+		/// conversions to `u128`, which are used in the computations.
+		type Balance: Default
+			+ AtLeast32BitUnsigned
+			+ CheckedAdd
+			+ CheckedDiv
+			+ CheckedMul
+			+ CheckedSub
+			+ Codec
+			+ Copy
+			+ MaxEncodedLen
+			+ Ord
+			+ Parameter
+			+ Zero;
+	}
+
+	// ----------------------------------------------------------------------------------------------------
+	//                                             Pallet Types
+	// ----------------------------------------------------------------------------------------------------
+
+	type BalanceOf<T> = <T as Config>::Balance;
+	type TimestampOf<T> = <T as Config>::Timestamp;
+	type VammIdOf<T> = <T as Config>::VammId;
+	type VammStateOf<T> = VammState<BalanceOf<T>, TimestampOf<T>>;
+
+	// ----------------------------------------------------------------------------------------------------
+	//                                           Runtime  Storage
+	// ----------------------------------------------------------------------------------------------------
+
+	/// The number of virtual markets, also used to generate the next market
+	/// identifier.
+	///
+	/// # Note
+	///
+	/// Frozen markets do not decrement the counter.
+	#[pallet::storage]
+	#[pallet::getter(fn vamm_count)]
+	#[allow(clippy::disallowed_types)]
+	pub type VammCounter<T: Config> = StorageValue<_, VammIdOf<T>, ValueQuery>;
+
+	/// Maps [VammId](Config::VammId) to the corresponding virtual
+	/// [VammState] specs
+	#[pallet::storage]
+	#[pallet::getter(fn get_vamm)]
+	pub type VammMap<T: Config> = StorageMap<_, Blake2_128Concat, VammIdOf<T>, VammStateOf<T>>;
+
+	// ----------------------------------------------------------------------------------------------------
+	//                                            Runtime Events
+	// ----------------------------------------------------------------------------------------------------
+
+	// Pallets use events to inform users when important changes are made.
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Emitted after a successful call to the [`create`](Pallet::create) function.
+		Created { vamm_id: VammIdOf<T>, state: VammStateOf<T> },
```

Comment:
Might be good to check with others how much information is reasonable to put in an event (maybe Substrate's stack exchange?)

Reply:
I'll have a look at it, but the `VammState` should be pretty small, only 3 `u128` and a `Option<i64>` as of now.

If size is a concern, we could remove the `state` tho.